src/torch_ucc.cpp

```diff
@@ -133,65 +128,63 @@ ProcessGroupUCC::WorkUCC::~WorkUCC() {
 #endif
 }
 
+void ProcessGroupUCC::WorkUCC::setException() {
+  if (exception()) {
+    return;
+  }
+  exception_ = entry_->eptr_;
+}
+
+void ProcessGroupUCC::WorkUCC::setAndThrowException() {
+  setException();
+  if (exception()) {
+    std::rethrow_exception(exception());
+  }
+}
+
 bool ProcessGroupUCC::WorkUCC::isCompleted() {
-  return ((status_ == UCC_OK) || (status_ < 0));
+  setException();
+  return exception() || entry_->status_ <= 0;
 }
 
 bool ProcessGroupUCC::WorkUCC::isSuccess() const {
-  return (status_ >= 0);
+  return !exception() && entry_->status_ == 0;
 }
 
 bool ProcessGroupUCC::WorkUCC::wait(std::chrono::milliseconds /* unused */) {
 #ifdef USE_CUDA
   if (fence && !torch_ucc_config.blocking_wait[(int)opType_]) {
```

Comment:
not relate to this PR, just a question of a corner case comes to mind now - what if a request is progressed and completed before main thread calls wait (e.g., small collective)? it seems no point to block the stream anyhow here in such scenario and we may save a few micro/nano-seconds?

Reply:
I agree, moreover we can check that collective completes immediately in enqueue_collective or enqueue_cuda_collective and do not even bother progress thread in that case. Maybe create a separate PR for this change?